### PR TITLE
[MRG] Implement iterative reweighted TF-MxNE

### PR DESCRIPTION
**Pull request by mathurinm**
Currently in progress as I still need to adapt the dual scaling computation, and also the w_time computation. I'll give it a try this week.
@agramfort @joewalter

### CODE REVIEW

**Comment by mathurinm** [comment]
I think it is working. I have added a IR case in the tfmxne example, but I'm struggling to find values of alpha/l1_ratio which show a benefit (less biased dipoles). In fact, it seems worse for the range of values I've manually tried.

**Comment by mathurinm** [comment]
@agramfort @joewalter I'm done but the new example on somato is not very convincing. Also what should I test to assert solver convergence, since we can't rely on duality gap ?

File: examples/inverse/plot_multidict_reweighted_tfmxne.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+    n_tfmxne_iter=n_tfmxne_iter, loose=loose,
+    depth=depth,
+    wsize=wsize, tstep=tstep, return_as_dipoles=True,
+    return_residual=True)
+
+# Crop to remove edges
+for dip in dipoles:
+    dip.crop(tmin=0., tmax=0.3)
+evoked.crop(tmin=0., tmax=0.3)
+residual.crop(tmin=0., tmax=0.3)
```
Comment (agramfort) [inline examples/inverse/plot_multidict_reweighted_tfmxne.py]: I like to see the baselines in the plots. You should see that dipoles amplitudes are zero during baseline

**Comment by agramfort** [comment]
typically 10 reweighting is already an upper bound

see how we do it with irMxNE implemen and let's be consistent.

**Comment by larsoner** [comment]
@mathurinm @joewalter this one is almost a year old, what's the plan?

**Comment by mathurinm** [comment]
I had trouble finding a convincing example to put in the gallery: performance was quite sensitive to cropping tmin for example.
I am a bit busy with other things for ~6 weeks but I will get back on in May.

File: examples/inverse/plot_mixed_norm_inverse.py [hunk 2]
```diff
@@ -106,9 +106,9 @@
 src_fsaverage_fname = subjects_dir + '/fsaverage/bem/fsaverage-ico-5-src.fif'
 src_fsaverage = mne.read_source_spaces(src_fsaverage_fname)
 
-plot_sparse_source_estimates(src_fsaverage, stc_fsaverage, bgcolor=(1, 1, 1),
-                             fig_name="Morphed %s (cond %s)" % (solver,
-                             condition), opacity=0.1)
+plot_sparse_source_estimates(
```
Comment (mathurinm) [inline examples/inverse/plot_mixed_norm_inverse.py]: @agramfort we don't break lines right after function call in MNE, do we?

**Comment by larsoner** [comment]
@mathurinm you need a rebase

**Comment by mathurinm** [comment]
Here's what I obtain on my somato example. 

image
image

The sfreq is 300 Hz which is not high enough to see very transient bursts as in @yousrabk and @joewalter paper. We have the MIND-dataset on which the results are prettier,  (last plot) maybe we can merge this one first, package the new datasets into MNE and update the example then ?

image

File: examples/inverse/plot_multidict_reweighted_tfmxne.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+
+###############################################################################
+# Generate stc from dipoles
+stc = make_stc_from_dipoles(dipoles, forward['src'])
+
+plot_sparse_source_estimates(forward['src'], stc, bgcolor=(1, 1, 1),
+                             opacity=0.1, fig_name="irTF-MxNE (cond %s)"
+                             % condition)
+# modes=['sphere'], scale_factors=[1.])
```
Comment (agramfort) [inline examples/inverse/plot_multidict_reweighted_tfmxne.py]: cleanup

File: examples/inverse/plot_multidict_reweighted_tfmxne.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+evoked.crop(tmin=0.008, tmax=0.2)
+
+# Compute noise covariance matrix
+cov = mne.compute_covariance(epochs, rank='info', tmax=0.)
+
+# Handling forward solution
+forward = mne.read_forward_solution(fwd_fname)
+
+###############################################################################
+# Run iterative reweighted multidict TFMxNE solver
```
Comment (agramfort) [inline examples/inverse/plot_multidict_reweighted_tfmxne.py]: TFMxNE -> TF-fMxNE

please check everywhere.

File: mne/inverse_sparse/mxne_optim.py [hunk 5]
```diff
@@ -1138,26 +1215,22 @@ def _tf_mixed_norm_solver_bcd_(M, G, Z, active_set, candidates, alpha_space,
...
+                        i + 1, np.sum(active_set) / n_orient))
             logger.info("    dgap %.2e :: p_obj %f :: d_obj %f" % (
                         gap, p_obj, d_obj))
 
         if converged:
             break
 
-        if (ii == maxit - 1):
-            converged = False
-            break
```
Comment (agramfort) [inline mne/inverse_sparse/mxne_optim.py]: using for else instead?

**Comment by mathurinm** [comment]
The involved math derivations are detailed in a pdf that I've uploaded to my webpage for potential future needs:
https://github.com/mathurinm/mathurinm.github.io/commit/6d159e304004354d5425a29ce4f472d130074861

**Comment by agramfort** [comment]
thx heaps @mathurinm !

**Comment by agramfort** [comment]
@mathurinm oups I forgot to ask you to update what's new (latests.inc) page? can you do a quick PR ? thx

**Pull request merged**